ARCs/arc-0014.md

```diff
@@ -0,0 +1,246 @@
+> Comments like this are non-normative.
+
+Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects, and all JSON string types are UTF-8 encoded.
+
+This ARC uses interchangeably the terms "*blockchain address*", "*public key*", and "*PK*" to indicate the on-chain address of a blockchain identity, and in particular of an Algorand account.
+
+### Overview
+
+This document describes a standard approach to authenticate Users with their blockchain identity. In particular, in this ARC, Algorand address are used as a *unique identifiers*, and the secret key of that address is used to digitally sign a *message* as a proof of identity for a  Verifier.
+
+To sum up, given an Algorand account <*PKa, SKa*>, this ARC defines the standards for:
+
+- create an [ARC-14](./arc-0014.md) compliant digital signature with *SKa*;
+- verify an ARC-14 compliant digital signature with *PKa*.
+
+### Assumptions
+
+The standard proposed in this document works under the following assumptions:
+
+- User and Verifier communicates over secure SSL/TLS encrypted channels;
+- The Verifier knows the Users’ *PKa*;
+- For each *PKa* the Verifier generates a unique message to be signed;
+- The message changes for each authentication request - avoid replay attack;
+- Users are the only custodians of the *SKa*;
+- The secret keys are stored into a wallet;
+- Users won’t use different *PKa* for authentication;
+- Users won’t change their public address for authentication;
+- Users **MUST** use Algorand compliant keys to sign the messages;
+- Multisig and LogicSig are not supported;
+- Users **MUST NOT** use rekeyed accounts.
+
+> Rekey breaks ARC-14 security. Applications with high security requirements **MUST** check that Users' accounts have not been rekeyed.
+
+### Authentication Mechanism
+
+The authentication mechanism defined in this ARC works as follows: a User communicates to the Verifier her intention to authenticate with her Algorand account <*PKa, SKa*>.
+
+> Algorand transforms traditional 32-bytes cryptographic keys into more readable and user-friendly objects. A detailed description of such a transformation can be found in the <a href="https://developer.algorand.org/docs/get-details/accounts/#keys-and-addresses">developer portal</a>.
+
+The Verifier requests the signing of a random message is required to be signed with the secret key *SKa* associated with the User's Algorand account. The User generates the digital signature using a Wallet, and sends it back to the Verifier. At that stage, the Wallet **MUST** check that the User is signing the message for the correct Verifier (to protect Users against man-in-the-middle attack). Finally, the Verifier verifies the signature and, if everything is good, authenticates the User.
```

Comment:
Shall we elaborate on the mechanism for the wallet to verify that msg comes from the correct Verifier?

Reply:
One solution is to require the `service` in the `AuthMessage` to be a domain name and to require the wallet to check that the domain in the `AuthMessage` matches the domain that was asked to use for the transaction to be signed.

This way, a man-in-the-middle attack could be more easily prevented.

```mermaid
sequenceDiagram
	participant User 
    participant Malicious Frontend  
    participant Wallet
    participant Malicious Backend
    participant Verifier
    User-->>Malicious Frontend: Connect to User's Wallet with PKa
    Malicious Frontend->>Wallet: Connect to Malicious Backend with PKa
	Wallet->>Malicious Backend: Get auth msg from Malicious Backend for PKa
	Malicious Backend->>Verifier: Get msg for PKa
	Verifier->>Verifier: Create a new msg for PKa with nonce
	Verifier->>Verifier: Store msg for later verification
	Verifier->>Malicious Backend: msg({service: 'verifier_domain'})
	Malicious Backend->>Wallet: msg({service: 'verifier_domain'})
	User-->>Wallet: Sign 0 fee Txn / auth msg
	Wallet->>Wallet: Check <msg.service, Malicious Backend domain>
alt  domain check fails
    Wallet->>Malicious Frontend: Authentication failure
else  domain check succeds or no domain check
	Wallet->>Malicious Backend: <PKa, Sig(msg)>
	Malicious Backend->>Verifier: Verify Sig(msg) with PKa and msg
	Verifier->>Wallet: Return authentication response (e.g. session JWT)
	Wallet->>Malicious Frontend: Return authentication response (e.g. session JWT)
end
```

Reply:
Hi, i can speak for original arc0014 .. this is how it is supposed to work:

You can check for example how this process is implemented (to make account online):
https://www.youtube.com/watch?v=cSJjmnzvvLY

Source code of a-wallet: 
https://github.com/scholtz/wallet/blob/1003f2ca1e2e8f76b264d97817b631ae3ccd7ed8/src/store/kmd.js#L19
KMD:
https://github.com/scholtz/AlgorandKMDServer/blob/f9d04b717f0f58cf9151bd8fa9a65b7e805db87c/Controllers/KMDController.cs#L122
swagger: https://kmd.h2.a-wallet.net/swagger/
AlgorandAuthenticationDotNet:
https://github.com/scholtz/AlgorandAuthenticationDotNet/blob/main/AlgorandAuthenticationHandler.cs

@startuml

User -> Wallet: User wish to login to the website .. Note as realm, params as session timeout
Wallet -> User: Signed tx

User -> Website: Send auth message, web stores the message as auth token
Website --> User: Displays eg auth address

Website -> Backend: Request authorized method, use SigTx Authorization header

Backend -> AnyAlgorandNode : Request /v2/transactions/params
AnyAlgorandNode  -> Backend : Returns current block round number

Backend -> Website: 401 if signature, genesis or round is wrong
Backend -> Website: 200 if authorized and returns content

@enduml

![image](https://user-images.githubusercontent.com/1223439/195992722-c1df411a-4f7e-4c00-ac7b-46c968dfd1cd.png)

Reply:
> > 
> 
> @SudoWeezy Sorry, I am a bit confused now. We have two open ARC-14 PRs at the moment, this one (#84), and the previous one (#41).
> 
> We have a number of comments on both and both seem to have gathered some consensus. What is the standardization process we should follow to move on?
> 
> cc @jannotti

I recommend to rename the later PR to ARC-0041, and keep original ARC-0014 as it was proposed.

Reply:
Hi @lorenzopolidori. Thank you for taking time to review this ARC. The `service` field of the `AuthMessage` is intended meant to include the Verifier domain name. Wallets can easily check that the domain name of the requestor matches the one in the Authentication Message.

We can definitely elaborate more on the description and diagram. I won't introduce other components in the diagram though, e.g. `Frontend` or `Backend`. We should keep things as generic as possible. WDYT?

Reply:
Hi @deanstef, thanks for putting forward this ARC. I would not necessarily add the man-in-the-middle attack example in the ARC, however I think that it would be important to explain more in detail some security aspects in the ARC (e.g. the domain name check), as well as  have a number of help items published together with the ARC once it becomes a standard for the developers to better understand it, e.g. a demo app (as others suggested) and some extra documentation on how the ARC mitigates security risks.

Reply:
@deanstef why are you ignoring me? If you do not want to cooperate on my ARC, why do you have need to hijack my ARC number?

regarding your  Frontend vs Backend .. there is no way users can do security around the Frontend.. user may just see which account he is using.. the real security must come at the Backend when user is writing some data or reading sensitive data.. There is no point to be generic at this point as it has two different security scopes..

Reply:
Hi @scholtz, I am not ignoring you. I'd like to collaborate with anyone which is proactive and stay focussed on the scope of ARC14.

Please note that ARCs define standards, not implementations. We should keep a standard as much generic as possible and leave users the implementation choices.

Also, let me clarify something: this ARC came after the analysis of your initial proposal (thank you for putting things together). It is meant to handle the **authentication** of users via Algorand account. It was called "_alternative ARC0014_". I do not care about the ARC number, it is something that @SudoWeezy manages. However, I think we should keep things as they are now, because several projects are already referencing ARC14 with the specs here defined.

We decided to keep Authentication and Session Handling in two separate ARCs. Since your comments keep showing sessions components, which are out of scope here, I would suggest you to work on the correct ARC (PR #41). Thank you !

Reply:
@lorenzopolidori agreed. I'll do a simple demo app, and make more clear the domain verification part for Wallets. TY

Reply:
Just to clarify some things:
- Everything that is not in the **main** repository with the status **final** can change.
- **Whoever submits it**, every ARC must be approved by the majority of the community.

**If** a consensus is reached on this PR:
@scholtz, your [Initial PR](https://github.com/algorandfoundation/ARCs/pull/41), which handles the Session, will need to require this version ARC-14 and start from there. It will then have a new number > 14.

Reply:
@scholtz I had a look at your demo projects and all i see is Under Construction, but more importantly in the EU you cannot offer perosnal PK auth without GDPR compliance, which your websites don't offer. The entire Arc14 and signed token discussion here is totally lacking examples, use cases, discussions around privacy, complications like rekeying and multisig, and parallels with other developments and trends around passkeys and passwordless authentication.  While I understand your desire to defend something you initiated from what you perceive to be unwanted change, what you are witnessing here is a process where people are collaboratively evolving an idea. An idea, which I think will ultimately be binned or evolved beyond recognition anyway.

Reply:
> @scholtz sorry, i do not offer PK.. it is the role of the wallet. Implementation of ARC14 should not require the GDPR compliance as you can technically have the private keys stored for example in azure vault or user may just have the mnemonics. I dont think that the technical guidelines how to do security should force you how to do legal bindings on how to publish or process the public key of the user. Regarding GDPR compliance and my projects, i believe that you are obliged to inform about GDPR only if you process personal data. You process personal data only if you store at least two personal informations classified by GDPR, eg user name and his public key. The wallet does not store any name, and the keys are stored purly in the user browser database. I will store sensitive data with the DREM project, and when the project will run, i am sure i will comply with GDPR rules. (I have extensive knowledge about gdpr because i created medical project which stored highly sensitive data which involved personal numbers and medical records (for covid testing))

complications like rekeying and multisig may be solved only after update in algod... lets start this discussion.. i want to know which account is rekeyed to which account and I want to have this information accessible by API.. if we solve this, we do not have issue with multisig.. the algod has this information and should not be too difficult to implement it.. @deanstef your profile seems to be from the algorand inc.. can you give us little light into the issue if this may be implemented in algod?

i am in favor of "collaboratively evolving an idea" .. however i do not want to see that some guys are putting there breaking changes and suggesting me that i should start new arc for the thing i proposed.. if the idea should be changed beyound the recognition of the original idea, i believe it should have different arc number

Reply:
That is incorrect. An authentication standard is useless without a registration step and a wallet public key registered with the server is personal information according to GDPR laws if the key is shared across relying parties.

Reply:
@scholtz You can already see if an account is rekeyed via `GET /v2/accounts/{address}`. If the account is rekeyed, the response will have a `auth-addr` field

https://developer.algorand.org/docs/rest-apis/algod/v2/#account

Reply:
Thank you @joe-p. 

@scholtz You can get the same `account info` via any Algorand SDK, e.g.: https://py-algorand-sdk.readthedocs.io/en/latest/algosdk/v2client/algod.html#algosdk.v2client.algod.AlgodClient.account_info

Reply:
@FrankSzendzielarz Hi. I am not sure about GDPR compliance here (not an expert). Do you have any comments/suggestions on that matter for ARC14?